### PR TITLE
Modernise GitCloud and update to latest best practices.

### PR DESCRIPTION
**Pull request by samscott89**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,112 +1,34 @@
-# GitCloud
```
Comment (samscott89) [inline README.md:1]: Revert this change + update so we have a nice readme.

File: services/gitclub/app/routes/helpers.py [hunk 2]
```diff
@@ -1,24 +1,97 @@
 from os import getenv
```
Comment (samscott89) [inline services/gitclub/app/routes/helpers.py:1]: Rename this to "authorization.py" and comment it thoroughly -- we'll probably want people to follow this a bunch.

File: frontend/api/roleChoices.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import useUser from "../lib/useUser";
+import { get } from "./common";
+
+const org = () => {
+    // const { userId } = useUser();
```
Comment (killpack) [inline frontend/api/roleChoices.ts:5]:
```suggestion
```

File: frontend/api/roleChoices.ts [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+    return get(`/org_role_choices`) as Promise<string[]>
+};
+
+const repo = () => {
+    // const { userId } = useUser();
```
Comment (killpack) [inline frontend/api/roleChoices.ts:10]:
```suggestion
```

File: frontend/components/OrganizationList.tsx [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+export function OrganizationList({ organizations }: { organizations: Org[] }) {
+  const [orgs, setOrgs] = useState<Org[]>(organizations)
+  const api = orgApi();
+
+  useEffect(() => {
+    const updateOrgs = async () => {
+      for (const org of orgs) {
+        const userCount = await api.userCount("" + org.id);
+        console.log(`Got user count ${userCount} for org ${org.id}`)
```
Comment (killpack) [inline frontend/components/OrganizationList.tsx:18]:
```suggestion
```

File: frontend/components/OrganizationList.tsx [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+                <div className="px-4 py-4 sm:px-6">
+                  <div className="flex items-center justify-between">
+                    <p className="text-sm font-medium text-indigo-600 truncate">
+                      {org.name}
+                    </p>
+                    {/* <div className="ml-2 flex-shrink-0 flex">
+                      <p className="px-2 inline-flex text-xs leading-5 font-semibold rounded-full bg-green-100 text-green-800">
+                        Foo
+                      </p>
+                    </div> */}
```
Comment (killpack) [inline frontend/components/OrganizationList.tsx:43]:
```suggestion
```

File: frontend/pages/login.tsx [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+  const router = useRouter();
+  const [login, setLogin] = useState<string>("");
+
+  const api = sessionApi();
+
+  const [error, setError] = useState<Error | undefined>(undefined);
+  if (error) return <ErrorPage error={error} setError={setError} />
+
+  // If a logged-in user navigates to this page, redirect to home.
+  // if (currentUser.isLoggedIn) router.push("/");
```
Comment (killpack) [inline frontend/pages/login.tsx:19]:
```suggestion
```

File: frontend/pages/login.tsx [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+  async function handleSubmit(e: FormEvent) {
+    e.preventDefault();
+    try {
+      setUsername(login);
+      // const u = await api.login({ username: login });
+      // mutateUser(u);
```
Comment (killpack) [inline frontend/pages/login.tsx:25]:
```suggestion
```

File: frontend/pages/logout.tsx [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    const [error, setError] = useState<Error | undefined>(undefined);
+
+    const api = sessionApi();
+
+    // If a logged-out user navigates to this page, redirect to home.
+    useEffect(() => {
+        api.logout().then(() => {
+            setUsername("");
+            router.replace(`/`);
+            // mutateUser(undefined)
```
Comment (killpack) [inline frontend/pages/logout.tsx:20]:
```suggestion
```

File: frontend/pages/orgs/[orgId].tsx [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+  } = orgApi().show(orgId);
+  const {
+    data: repos,
+    isLoading: repoLoading,
+    error: repoError,
+  } = repoApi(orgId).index();
+  const [roleChoices, setRoleChoices] = useState<string[]>([]);
+  useEffect(() => {
+    roleChoicesApi.org().then(setRoleChoices);
+    // .catch((e) => error(`Failed to fetch org role choices: ${e.message}`));
```
Comment (killpack) [inline frontend/pages/orgs/[orgId].tsx:34]:
```suggestion
```

File: frontend/pages/orgs/[orgId]/repos/[repoId]/settings.tsx [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+  const {
+    data: roleAssignments,
+    isLoading: rolesLoading,
+    error: rolesError,
+    mutate: setAssignments
+  } = roleAssignmentsApi.repo(orgId, repoId).index();
+
+  useEffect(() => {
+    roleChoicesApi.repo().then(setRoleChoices);
+    // .catch((e) => error(`Failed to fetch repo role choices: ${e.message}`));
```
Comment (killpack) [inline frontend/pages/orgs/[orgId]/repos/[repoId]/settings.tsx:44]:
```suggestion
```

File: frontend/pages/orgs/[orgId]/repos/[repoId]/settings.tsx [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+
+  function update(user: User, role: string) {
+    api.update({ username: user.username, role }).then((next) => {
+      const { username } = next.user;
+      // NOTE(gj): Assumes a user has a single role per resource.
+      setAssignments((as) =>
+        as?.map((a) => (a.user.username === username ? next : a))
+      );
+    }).catch(setError);
+    // .catch((e) => error(`Failed to update role assignment: ${e.message}`));
```
Comment (killpack) [inline frontend/pages/orgs/[orgId]/repos/[repoId]/settings.tsx:71]:
```suggestion
```

File: frontend/tsconfig.json [hunk 13]
```diff
@@ -1,26 +1,20 @@
 {
   "compilerOptions": {
-    "target": "es6",
-    "lib": [
-      "dom",
-      "dom.iterable",
-      "esnext"
-    ],
+    "target": "es5",
```
Comment (killpack) [inline frontend/tsconfig.json:3]: Any reason not to emit es6...?

File: policy/authorization.polar [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+
+# resource File { 
+#     permissions = ["read", "write"];
+#     relations = { folder: Folder };
+
+#     "read"  if "reader" on "folder";
+#     "write" if "writer"  on "folder";
+#     "read"  if "write";
+# }
+
```
Comment (killpack) [inline policy/authorization.polar:155]: Any reason to keep these around...?

```suggestion
```

File: services/gitclub/app/__init__.py [hunk 15]
```diff
@@ -1,29 +1,45 @@
...
+            poolclass=StaticPool,
+        )
 
     # Init Flask app.
     app = Flask(__name__)
+    app.config["SESSION_COOKIE_SECURE"] = True
+    app.config["SESSION_COOKIE_SAMESITE"] = "None"
+    app.config["SESSION_COOKIE_HTTPONLY"] = True
+    app.config["PERMANENT_SESSION_LIFETIME"] = timedelta(1)
+    # app.config["SESSION_COOKIE_DOMAIN"] = "gitcloud-gitclub.fly.dev"if PRODUCTION else False
```
Comment (killpack) [inline services/gitclub/app/__init__.py:39]:
```suggestion
```

File: services/gitclub/app/fixtures.py [hunk 16]
```diff
@@ -73,36 +158,81 @@ def load_fixture_data(session):
...
     for repo in repos:
-        oso.tell("has_relation", repo, "parent", repo.org)
+        facts.append(["has_relation", { "type": "Repository", "id": str(repo.id)}, "organization", { "type": "Organization", "id": str(repo.org.id) }])
+        facts.append(["is_protected", { "type": "Repository", "id": str(repo.id)}, { "type": "Boolean", "id": str(repo.protected).lower()}])
+        if repo.public:
+            facts.append(["is_public", { "type": "Repository", "id": str(repo.id)}])
 
-    for issue in issues:
-        oso.tell("has_relation", issue, "parent", issue.repo)
+    # for issue in issues:
```
Comment (killpack) [inline services/gitclub/app/fixtures.py:166]: This whole section is a TODO for me, I believe.

File: services/gitclub/app/routes/helpers.py [hunk 17]
```diff
@@ -1,24 +1,97 @@
...
+
+def current_user():
+    if g.current_user is None:
+        raise Unauthorized
+    return object_to_typed_id(g.current_user)
 
 
+def tell(predicate: str, *args: Any):
+    print(f'oso-cloud tell {predicate} {",".join([str(a) for a in args])}')
+    return oso.tell(predicate, *[object_to_typed_id(a, True) for a in args])
```
Comment (killpack) [inline services/gitclub/app/routes/helpers.py:41]:
```suggestion
    return oso.tell(predicate, *[object_to_typed_id(a) for a in args])
```
I don't think we want to allow unbound things in `tell` commands.

File: services/gitclub/app/routes/issues.py [hunk 18]
```diff
@@ -1,64 +1,81 @@
...
     payload = request.get_json(force=True)
-    repo = g.session.get_or_404(Repo, id=repo_id)
+    repo = g.session.get_or_404(Repository, id=repo_id)
     if not authorize("create_issues", repo):
         raise NotFound
-    issue = Issue(title=payload["title"], repo=repo, creator_id=g.current_user.id)
+    issue = Issue(title=payload["title"], repo=repo, creator_id=g.current_user.username)
     g.session.add(issue)
     g.session.commit()
     oso.bulk_tell(
```
Comment (killpack) [inline services/gitclub/app/routes/issues.py:44]: I'm probably going to add a wrapper around this function, so that everything goes through `helpers` (soon to be called `authorization`).

File: services/gitclub/app/routes/orgs.py [hunk 19]
```diff
@@ -1,38 +1,62 @@
...
+    if g.session.query(Organization).filter(Organization.name==payload["name"]).first() is not None:
+        return "Organization with that name already exists", 400
+    org = Organization(**payload)
+    if not authorize("create", "Organization"):
         raise Forbidden
     g.session.add(org)
-    g.session.commit()
-    oso.tell("has_role", g.current_user, "owner", org)
-    return org.repr(), 201
+    g.session.commit()
```
Comment (killpack) [inline services/gitclub/app/routes/orgs.py:30]:
```suggestion
    g.session.commit()
```

File: services/gitclub/app/routes/orgs.py [hunk 20]
```diff
@@ -1,38 +1,62 @@
...
+    return "deleted", 204
+
+@bp.route("/<int:org_id>/user_count", methods=["GET"])
+@cache.memoize()
+def user_count(org_id):
+    if not authorize("read", {"type": "Organization", "id": str(org_id)}):
         raise NotFound
-    return org.repr()
+    org_users = get("has_role", { "type": "User",  }, {}, { "type": "Organization", "id": str(org_id) })
+    print("org users: ", org_users)
```
Comment (killpack) [inline services/gitclub/app/routes/orgs.py:61]:
```suggestion
```

File: services/gitclub/app/routes/users.py [hunk 21]
```diff
@@ -1,29 +1,48 @@
...
+
+@bp.route("/<username>/orgs", methods=["GET"])
+def org_index(username):
+    if not authorize("read_profile", { "type": "User", "id": username }):
+        raise NotFound
+
+    # get all the repositories that the user has a role for
+    orgs = query("has_role", { "type": "User", "id": username }, {}, { "type": "Organization"})
+    orgIds = list(map(lambda fact: fact[3].get("id", "_"), orgs))
+    print(orgs, orgIds)
```
Comment (killpack) [inline services/gitclub/app/routes/users.py:42]:
```suggestion
```

File: services/jobs/package.json [hunk 22]
```diff
@@ -0,0 +1,36 @@
+{
+  "name": "actions-service",
```
Comment (killpack) [inline services/jobs/package.json:2]: Worth calling this "jobs-service" or too fiddly?

File: services/jobs/src/index.ts [hunk 23]
```diff
@@ -23,14 +20,24 @@ declare module "express" {
...
 }
 
+const config =
+  process.env.PRODUCTION == "1"
+    ? {
+      db: pgDataSource,
+      frontend: "https://gitcloud.vercel.app",
+    }
+    : {
+      // db: localDataSource,
```
Comment (killpack) [inline services/jobs/src/index.ts:30]:
```suggestion
```

File: services/gitclub/app/fixtures.py (hunk 16 above)
Comment (samscott89) [reply services/gitclub/app/fixtures.py:166]: Nah, nothing to do here.
